cpp/path_module/algorithm/path.cpp

```diff
@@ -454,34 +465,28 @@ void Path::SubgraphNodes(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *
 }
 
 void Path::SubgraphAll(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory) {
-  mgp::memory = memory;
+  mgp::MemoryDispatcherGuard guard{memory};
   const auto arguments = mgp::List(args);
   const auto graph = mgp::Graph(memgraph_graph);
   const auto record_factory = mgp::RecordFactory(result);
   try {
     auto config = arguments[1].ValueMap();
-    SetConfig(config);
+    PathHelper path_helper{config};
+    PathSubgraph path_subgraph{path_helper, record_factory, graph};
 
-    std::unordered_set<mgp::Node> start_nodes;
-    if (arguments[0].IsList()) {
-      for (const auto element : arguments[0].ValueList()) {
-        GetStartNodes(element, graph, start_nodes);
-      }
+    auto start_value = arguments[0];
+    if (!start_value.IsList()) {
+      path_subgraph.Parse(start_value);
     } else {
-      GetStartNodes(arguments[0], graph, start_nodes);
+      for (const auto &list_item : start_value.ValueList()) {
+        path_subgraph.Parse(list_item);
+      }
     }
 
-    LabelSets labelFilterSets;
-    ParseLabels(config.At("labelFilter").ValueList(), labelFilterSets);
-
-    std::map<mgp::Node, std::int64_t> visited_nodes;
-    mgp::List to_be_returned_nodes;
-    for (const auto node : start_nodes) {
-      VisitNode(node, visited_nodes, true, config, 0, labelFilterSets, to_be_returned_nodes);
-    }
+    auto to_be_returned_nodes = path_subgraph.BFS();
 
     std::unordered_set<mgp::Node> to_be_returned_nodes_searchable;
-    for (auto node : to_be_returned_nodes) {
+    for (const auto &node : to_be_returned_nodes) {
       to_be_returned_nodes_searchable.insert(node.ValueNode());
     }
```

Comment:
Not possible because node is of type mgp::Value, and we need to turn it into mgp::Node with ValueNode(). I can do this which makes me feel smart but I don't know if it's better:

`std::transform(
        to_be_returned_nodes.begin(), to_be_returned_nodes.end(),
        std::inserter(to_be_returned_nodes_searchable, to_be_returned_nodes_searchable.begin()),
        [](const mgp::Value &node) { return node.ValueNode(); }
    );`

Reply:
This is good, do that yes :smile: